### PR TITLE
fixed issue that didn't allow a user to update a document when deleting elements in an array attribute

### PR DESCRIPTION
**Pull request by safwanyp**
Use `symmetricDifference` instead of `difference`.
- `difference` only returns values that are present in the edited attributes, but not in the originally fetched attribute vales. This always ends up returning an empty array.
- I used `symmetricDifference` as it returns an intersection of `difference` from both sides and works perfectly for this use case, as the original attributes never change.

## What does this PR do?

This PR closes https://github.com/appwrite/appwrite/issues/5476

## Test Plan

- In addition to the existing tests, I performed manual testing by adding and deleting elements in an array attribute. 
- I was going to change the way the `difference` function worked, but realized it was being used in other places, which broke things. I used the `symmetricDifference` function as it fits in perfectly for this use case.

## Related PRs and Issues

* https://github.com/appwrite/appwrite/issues/5476

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes ✅

### CODE REVIEW

**Comment by safwanyp** [comment]
Okay I just realized there's another bug. 
Although elements can be deleted, existing items cannot be edited. The `Update` button stays disabled even after changing values. Should I create a new issue for that, or push a commit to fix it in this PR?

**Comment by safwanyp** [comment]
Okay so I figured out the issue that prevents updating of a document when editing the elements of an array attribute. 

In the code it seems like `$work` stores the value of edited values. `$work` is populated by iterating over `$doc`, but they are referencing the same underlying value, no? Due to this, when `$work` is populated, `$doc` is updated as well, which results in both the arrays having no difference. 

I can think of 2 solutions to this: 
1. Use `JSON.parse(JSON.stringify($doc))` to clone the $doc, and not reference copy.
2. Use a locally decalred variable instead of `formValues[attribute.key][index]` in `<AttributeItem />`, which can be passed into the `<Attribute />` component. Obviously, the local variable will be updated when there's a change.

I have only tested with the first approach as it doesn't mess with the functionality outside of the required component.

**Comment by TGlide** [comment]
> Okay so I figured out the issue that prevents updating of a document when editing the elements of an array attribute.
> 
> In the code it seems like `$work` stores the value of edited values. `$work` is populated by iterating over `$doc`, but they are referencing the same underlying value, no? Due to this, when `$work` is populated, `$doc` is updated as well, which results in both the arrays having no difference.
> 
> I can think of 2 solutions to this:
> 
>     1. Use `JSON.parse(JSON.stringify($doc))` to clone the $doc, and not reference copy.
> 
>     2. Use a locally decalred variable instead of `formValues[attribute.key][index]` in `<AttributeItem />`, which can be passed into the `<Attribute />` component. Obviously, the local variable will be updated when there's a change.
> 
> 
> I have only tested with the first approach as it doesn't mess with the functionality outside of the required component.

Hi @safwanyp! Thank you for the PR 😄 

I'd say doing a deepClone with JSON.parse should be enough. I'd recommend creating an utility for it in `helpers/object.ts`

**Comment by safwanyp** [comment]
Hey @TGlide, thanks for the feedback and pointers!

I've just pushed a commit that creates a `deepClone` utility function, and uses it to create a copy of `$doc` into `$work`. Tests were successful on local copy. Appreciate any feedback!

File: src/routes/console/project-[project]/databases/database-[database]/collection-[collection]/document-[document]/data/+page.svelte [hunk 1]
```diff
@@ -13,30 +13,33 @@
...
+                        '$databaseId',
+                        '$createdAt',
+                        '$updatedAt'
+                    ].includes(key);
+                })
+                .reduce((obj, key) => {
+                    obj[key] = $doc[key];
+                    return obj;
+                }, {}) as Models.Document
+        )
```
Comment (TGlide) [inline src/routes/console/project-[project]/databases/database-[database]/collection-[collection]/document-[document]/data/+page.svelte:42]: I think this function has become quite verbose.

As an alternative, we can create an initWork function, where we have better control flow.

e.g.

```ts
function initWork() {
  let prohibitedKeys = ['$id', ...];
  const filteredKeys = Object.keys($doc)
  // So on and so forth..
}

const work = initWork()
```

Comment (safwanyp) [reply src/routes/console/project-[project]/databases/database-[database]/collection-[collection]/document-[document]/data/+page.svelte:42]: that's fair. I'll work on this 👍🏻

**Review by TGlide — APPROVED** [review]

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**